gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/FSSpecStore.java

```diff
@@ -98,17 +101,40 @@ public FSSpecStore(Config sysConfig, SpecSerDe specSerDe, Optional<Logger> log)
     }
   }
 
+  /**
+   * @param specUri path of the spec
+   * @return empty string for topology spec, as topolgies do not have a group,
+   *         group name for flow spec
+   */
+  public static String getSpecGroup(Path specUri) {
+    return specUri.getParent().getName();
+  }
+
+  public static String getSpecName(Path specUri) {
+    return Files.getNameWithoutExtension(specUri.getName());
+  }
+
   @Override
   public boolean exists(URI specUri) throws IOException {
     Preconditions.checkArgument(null != specUri, "Spec URI should not be null");
-
-    FileStatus[] fileStatuses = fs.listStatus(this.fsSpecStoreDirPath);
+    Path flowPath = new Path(specUri.getPath());
+    String specGroup = getSpecGroup(flowPath);
+    String specName = getSpecName(flowPath);
+    FileStatus[] fileStatuses;
+    try {
+      if (StringUtils.isEmpty(specGroup)) {
+        fileStatuses = fs.listStatus(this.fsSpecStoreDirPath);
+      } else {
+        fileStatuses = fs.listStatus(new Path(this.fsSpecStoreDirPath, specGroup));
+      }
+    } catch (FileNotFoundException e) {
+      return false;
+    }
     for (FileStatus fileStatus : fileStatuses) {
-      if (StringUtils.startsWith(fileStatus.getPath().getName(), specUri.toString())) {
+      if (!fileStatus.isDirectory() && fileStatus.getPath().getName().startsWith(specName)) {
```

Comment:
We need to revisit having a version delimiter. Please add a note here. Currently without a delimiter the prefix check may match other specs that should not be matched.